src/neo/SmartContract/ApplicationEngine.Contract.cs

```diff
@@ -100,19 +100,19 @@ internal void DestroyContract()
                     Snapshot.Storages.Delete(key);
         }
 
-        internal void CallContract(UInt160 contractHash, string method, Array args)
+        internal void CallContract(UInt160 contractHash, ContractParameterType returnType, string method, Array args)
         {
-            CallContractInternal(contractHash, method, args, CallFlags.All);
+            CallContractInternal(contractHash, returnType, method, args, CallFlags.All);
         }
 
-        internal void CallContractEx(UInt160 contractHash, string method, Array args, CallFlags callFlags)
+        internal void CallContractEx(UInt160 contractHash, ContractParameterType returnType, string method, Array args, CallFlags callFlags)
         {
             if ((callFlags & ~CallFlags.All) != 0)
                 throw new ArgumentOutOfRangeException(nameof(callFlags));
-            CallContractInternal(contractHash, method, args, callFlags);
+            CallContractInternal(contractHash, returnType, method, args, callFlags);
         }
 
-        private void CallContractInternal(UInt160 contractHash, string method, Array args, CallFlags flags)
+        private void CallContractInternal(UInt160 contractHash, ContractParameterType returnType, string method, Array args, CallFlags flags)
```

Comment:
Why the caller need to specify the return type? it should be defined in the abi.

Reply:
For security reason. See the example:

```c#
bool verify(int a, int b)
{
    CallSomeContract();
    return a > b;
}
```

We expect `CallSomeContract()` to return no value. But if it maliciously returns 2 values, then these values will be automatically copied onto our stack, causing us to return wrong results.

```
//script of CallSomeContract()
PUSH 1 // this hide the parameter b of verify()
PUSH 2 // this hide the parameter a of verify()
RET
```

So we have to check the return values at the runtime.

Reply:
As `CallContract` and `CallContractEx` is to call contract public methods, we can read `returnType` from ABI directly. Why need user to pass the value?

Reply:
Because you can modify the abi at any time.

Reply:
we shluld not be able to modify the abi without change the script

Reply:
the ABI is apart from the script, no @shargon?

Reply:
Relate to https://github.com/neo-project/neo/pull/1577. I think the abi is the metadata of contract script, should not be modified, which is not a configurable item.

Reply:
> we shluld not be able to modify the abi without change the script

But you cannot stop him from modifying the ABI, because he can add a `NOP` to the script to make the script change.

Reply:
And even if the ABI cannot be modified, the attacker can upload the fake ABI at the first upload.

Reply:
> we shluld not be able to modify the abi without change the script

Attackers can modify ABI and Script at the same time. But we have dynamic calls, which becomes an attack point.

Reply:
If has an attack, the tx will execute failure, users will know and can check it.
```cs
if (rvcount != -1 && rvcount != context.EvaluationStack.Count)
   throw new InvalidOperationException();
```

Reply:
If a malicious can push more values than expected, also can send this values with the expected type. I think that if we need a specific type we can do it with `ISTYPE`

https://github.com/neo-project/neo-vm/blob/9d02720044d161c2a334d8e5a66130d3feda4066/src/neo-vm/ExecutionEngine.cs#L1166

Reply:
> If has an attack, the tx will execute failure, users will know and can check it.

It won't fail if the contract clean the stack before return.

```
INITSLOT 1, 0
//CallSomeContract();
GT
STLOC0
CLEAR
LDLOC0
RET
```

Reply:
@shargon We don't need to check the type. We just need to check the count of the values.

Reply:
I think that we only can prevent this issue with stack isolation. It's weird to me define the expected type by the caller. Maybe we shouldn't use clone in callbacks?

Reply:
Stack isolation cannot solve this problem. In fact, vm is already stack isolated. When a contract returns, the virtual machine automatically copies the return value to the caller's stack. We must ensure that the caller knows how many return values have been copied.

Reply:
Then maybe we need to recover the RVcount?

Reply:
The old `RVCount` in neo-vm also doesn't solve this problem. Because it reads the `RVCount` from ABI.

Reply:
But i don't understand why this solve the problem, if the call can `pop` anything and `push` everything, it can fake the result.

Reply:
No, it can't pop anything. The virtual machine automatically copies the return value to the caller's stack. So it can push anything.

Reply:
Maybe we can create a canary in this methods, store the number of items previously, and check it later, if it's more than expected (by compiler), we can throw an exception.

Reply:
Should we set `rvcount = 0` for `_initialize`, it may have the same issue.
https://github.com/neo-project/neo/blob/fc67879961e7e293e340adfded4ab6b4076e61d0/src/neo/SmartContract/ApplicationEngine.Contract.cs#L160-L161

Reply:
> if it's more than expected (by compiler)

How to do it?

> Should we set rvcount = 0 for _initialize, it may have the same issue.

`_initialize()` is not a cross-contract call, so it's okay.

Reply:
> But if it maliciously returns 2 values

Now the contract can only return one value, right?

Reply:
It is just an example. One value can also harm the system.